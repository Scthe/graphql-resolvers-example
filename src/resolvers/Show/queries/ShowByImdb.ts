import GqlContext from "GqlContext";
import { QueryShowByImdbArgs } from "typingsGql";
import { RootType as ShopType } from "../types/Show";

export default (
  _root: any,
  args: QueryShowByImdbArgs,
  context: GqlContext
): Promise<ShopType> => {
  // Here we do query to simplify implementation. We overfetch
  // if the user only wants the "imdbId" in response,
  // but it's MUCH simpler to write.
  const api = context.dataSources.showsAPI;
  return api.getByImdbId(args.imdbId);
};
