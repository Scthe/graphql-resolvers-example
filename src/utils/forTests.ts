import nock from "nock";

// utils used in tests

export const setupNock = (): void => {
  beforeEach(() => {
    // each jest test loads modules anew
    nock.disableNetConnect();
    nock.enableNetConnect("127.0.0.1");
  });

  afterEach(() => {
    nock.cleanAll();
  });

  afterAll(() => {
    nock.restore(); // https://github.com/nock/nock/issues/1817
  });
};
