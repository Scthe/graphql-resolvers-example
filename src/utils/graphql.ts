import GqlContext from "../GqlContext";

/** Resolver factory for simple properties */
export const copyFromRestResponse = <RootT, RestResponse, K extends keyof RestResponse>(
  getItem: (root: RootT, ctx: GqlContext) => Promise<RestResponse>,
  propName: K
) => {
  // we return resolver
  return async (
    root: RootT,
    _: any,
    context: GqlContext
  ): Promise<RestResponse[K]> => {
    const item = await getItem(root, context);
    return item[propName];
  }
};
