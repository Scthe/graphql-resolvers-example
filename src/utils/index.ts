export const parseId = (id: number | string): ID => {
  if (typeof id === "string") {
    return parseInt(id, 10) as any;
  }
  return id as any;
};
