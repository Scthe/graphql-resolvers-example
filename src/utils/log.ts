export default (...args: unknown[]): void => {
  if (process.env.NODE_ENV !== "test") {
    console.log(...args);
  }
};
