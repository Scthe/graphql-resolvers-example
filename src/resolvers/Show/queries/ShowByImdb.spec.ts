import supertest from "supertest";
import nock, { Scope } from "nock";
import merge from "deepmerge";

import app from "../../../index";
import * as gqlExamples from "./ShowByImdb.example";
import { Show as GqlSchemaShow } from "typingsGql";
import { setupNock } from "utils/forTests";

import ShowMock from "dataSources/tvmaze/mocks/show.mock.json";

describe("GraphQL: showByImdb()", () => {
  let scope: Scope | null = null;

  setupNock();

  afterEach(() => {
    scope && scope.done(); // verify all planned requests were actually done
  });

  const mockSuccessShowResponse = (imdbId: string) => {
    // NOTE: this allows for ONE request
    scope = nock("https://api.tvmaze.com")
      .get(`/lookup/shows?imdb=${imdbId}`)
      .reply(
        200,
        merge(ShowMock, {
          externals: {
            imdb: imdbId,
          },
        })
      );
  };

  test("should fetch basic item properties", () => {
    const imdbId = "tt3230854";
    mockSuccessShowResponse(imdbId);

    return supertest(app)
      .post("/graphql")
      .send({
        query: gqlExamples.GQL_SHOW_FIELDS,
      })
      .then((resp) => {
        const item = resp.body.data.showByIMDB as GqlSchemaShow;
        expect(item.id).toBe("" + ShowMock.id);
        expect(item.imdbId).toBe(imdbId);
        expect(item).toMatchSnapshot(); // just for a good measure
      })
      .finally();
  });
});
