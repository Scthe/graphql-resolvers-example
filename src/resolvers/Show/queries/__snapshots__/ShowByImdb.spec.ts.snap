// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL: showByImdb() should fetch basic item properties 1`] = `
Object {
  "genres": Array [
    "Drama",
    "Mystery",
    "Medical",
  ],
  "id": "118",
  "imdbId": "tt3230854",
  "language": "English",
  "name": "House",
  "premiered": "2004-11-16",
  "runtime": 60,
  "status": "Ended",
  "summary": "<p>Sink your teeth into meaty drama and intrigue with <b>House</b>, FOX's take on mystery, where the villain is a medical malady and the hero is an irreverent, controversial doctor who trusts no one, least of all his patients.</p>",
  "type": "Scripted",
}
`;
